gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/MysqlUserQuotaManager.java

```diff
@@ -47,55 +49,259 @@
+        return quotaCheck;
+      } else {
+        addDagId(connection, DagManagerUtils.generateDagId(dagNode).toString());
+      }
+
+      String proxyUser = ConfigUtils.getString(dagNode.getValue().getJobSpec().getConfig(), AzkabanProjectConfig.USER_TO_PROXY, null);
+      String flowGroup = ConfigUtils.getString(dagNode.getValue().getJobSpec().getConfig(),
+          ConfigurationKeys.FLOW_GROUP_KEY, "");
+      String specExecutorUri = DagManagerUtils.getSpecExecutorUri(dagNode);
+
+      boolean proxyUserCheck;
+
+      if (proxyUser != null && dagNode.getValue().getCurrentAttempts() <= 1) {
+        int proxyQuotaIncrement = incrementJobCountAndCheckQuota(connection,
+            DagManagerUtils.getUserQuotaKey(proxyUser, dagNode), getQuotaForUser(proxyUser), CountType.USER_COUNT);
+        proxyUserCheck = proxyQuotaIncrement >= 0;  // proxy user quota check succeeds
+        quotaCheck.setProxyUserCheck(proxyUserCheck);
+        if (!proxyUserCheck) {
+          // add 1 to proxyUserIncrement since proxyQuotaIncrement is the count before the increment
+          requesterMessage.append(String.format(
+              "Quota exceeded for proxy user %s on executor %s : quota=%s, requests above quota=%d%n",
+              proxyUser, specExecutorUri, getQuotaForUser(proxyUser), Math.abs(proxyQuotaIncrement) + 1 - getQuotaForUser(proxyUser)));
+        }
+      }
+
+      String serializedRequesters = DagManagerUtils.getSerializedRequesterList(dagNode);
+      boolean requesterCheck = true;
+
+      if (dagNode.getValue().getCurrentAttempts() <= 1) {
+        List<String> uniqueRequesters = DagManagerUtils.getDistinctUniqueRequesters(serializedRequesters);
+        for (String requester : uniqueRequesters) {
+          int userQuotaIncrement = incrementJobCountAndCheckQuota(connection, DagManagerUtils.getUserQuotaKey(requester, dagNode),
+              getQuotaForUser(requester), CountType.REQUESTER_COUNT);
+          boolean thisRequesterCheck = userQuotaIncrement >= 0;  // user quota check succeeds
+          requesterCheck = requesterCheck && thisRequesterCheck;
+          quotaCheck.setRequesterCheck(requesterCheck);
+          if (!thisRequesterCheck) {
+            requesterMessage.append(String.format("Quota exceeded for requester %s on executor %s : quota=%s, requests above quota=%d%n. ",
+                requester, specExecutorUri, getQuotaForUser(requester), Math.abs(userQuotaIncrement) + 1 - getQuotaForUser(requester)));
+          }
+        }
+      }
+
+    boolean flowGroupCheck;
+
+    if (dagNode.getValue().getCurrentAttempts() <= 1) {
+      int flowGroupQuotaIncrement = incrementJobCountAndCheckQuota(connection,
+          DagManagerUtils.getFlowGroupQuotaKey(flowGroup, dagNode), getQuotaForFlowGroup(flowGroup), CountType.FLOWGROUP_COUNT);
+      flowGroupCheck = flowGroupQuotaIncrement >= 0;
+      quotaCheck.setFlowGroupCheck(flowGroupCheck);
+      if (!flowGroupCheck) {
```

Comment:
For mysql quota, if we fail to check any of the quota type, shouldn't we directly rollback the change in this method?

Reply:
Do you mean if user check fails, should we immediately stop without trying to check proxy/requesterService quota? I think we can, but the existing code would check all the quotas anyway to be able to form a complete error message (in  `requesterMessage`) so I chose not to disturb that functionality

Reply:
I don't mean that, we still process to the end, but before commit the change, we check whether all three succeed, if not, instead of commit, we should revert the change at this point. But we still return all check result to user.